pkg/dockerregistry/server/repositorymiddleware.go

```diff
@@ -525,63 +468,87 @@ func (r *repository) getImage(dgst digest.Digest) (*imageapi.Image, error) {
 	return r.registryOSClient.Images().Get(dgst.String())
 }
 
-// getImageStreamTag retrieves the Image with tag `tag` for the ImageStream
-// associated with r.
-func (r *repository) getImageStreamTag(tag string) (*imageapi.ImageStreamTag, error) {
-	return r.registryOSClient.ImageStreamTags(r.namespace).Get(r.name, tag)
-}
-
 // getImageStreamImage retrieves the Image with digest `dgst` for the ImageStream
 // associated with r. This ensures the image belongs to the image stream.
 func (r *repository) getImageStreamImage(dgst digest.Digest) (*imageapi.ImageStreamImage, error) {
 	return r.registryOSClient.ImageStreamImages(r.namespace).Get(r.name, dgst.String())
 }
 
 // rememberLayers caches the provided layers
-func (r *repository) rememberLayers(manifest *schema1.SignedManifest, cacheName string) {
+func (r *repository) rememberLayers(manifest distribution.Manifest, cacheName string) {
 	if !r.pullthrough {
 		return
 	}
 	// remember the layers in the cache as an optimization to avoid searching all remote repositories
-	for _, layer := range manifest.FSLayers {
-		r.cachedLayers.RememberDigest(layer.BlobSum, cacheName)
+	for _, layer := range manifest.References() {
+		r.cachedLayers.RememberDigest(layer.Digest, cacheName)
 	}
 }
 
 // manifestFromImageWithCachedLayers loads the image and then caches any located layers
-func (r *repository) manifestFromImageWithCachedLayers(image *imageapi.Image, cacheName string) (*schema1.SignedManifest, error) {
-	manifest, err := r.manifestFromImage(image)
+func (r *repository) manifestFromImageWithCachedLayers(image *imageapi.Image, cacheName string, supportsSchema2 bool) (manifest distribution.Manifest, err error) {
+	if supportsSchema2 {
+		if image.DockerImageManifestMediaType == schema2.MediaTypeManifest {
+			manifest, err = r.deserializedManifestFromImage(image)
+		} else {
+			manifest, err = r.signedManifestFromImage(image)
+		}
+	} else {
+		if image.DockerImageManifestMediaType == schema2.MediaTypeManifest {
+			err = fmt.Errorf("unable to convert new image to old one")
+			err = regapi.ErrorCodeManifestInvalid.WithDetail(err)
```

Comment:
> In case of acceptschema2: true docker 1.8.2-fc22 gets Error: image tmp/busybox:latest not found

That's bad, it should work. See the 3rd paragraph in [upstream compatibility page under Manifest Push with Docker 1.10](https://github.com/docker/distribution/blob/master/docs/compatibility.md#manifest-push-with-docker-110).